fe/parameterize_batch.py

```diff
@@ -0,0 +1,487 @@
+                rot_matrix = special_ortho_group.rvs(3).astype(dtype=np.float64)
+                guest_conf = np.matmul(guest_conf, rot_matrix)*10
+
+                for atom_idx, pos in enumerate(guest_conf):
+                    conformer.SetAtomPosition(atom_idx, (float(pos[0]), float(pos[1]), float(pos[2])))
+
+            open_ff = forcefield.Forcefield(args.forcefield)
+            nrg_fns = open_ff.parameterize(guest_mol)
+            guest_masses = get_masses(guest_mol)
+            guest_system = system.System(nrg_fns, open_ff.params, open_ff.param_groups, guest_masses)
+
+            combined_system = host_system.merge(guest_system)
+            cbs = -1*np.ones_like(np.array(combined_system.masses))*0.0001
+            lambda_idxs = np.zeros(len(combined_system.masses), dtype=np.int32)
+            lambda_idxs[num_host_atoms:] = -1
+
+            sim = simulation.Simulation(
+                combined_system,
+                step_sizes,
+                cas,
+                cbs,
+                lambda_schedule,
+                lambda_idxs,
+                precision
+            )
+
+            epoch_train_params = get_params(opt_state)
+
+            sim.system.params = epoch_train_params
+
+            all_args = []
+
+            child_conns = []
+            parent_conns = []
+            for conf_idx in range(num_conformers):
+
+                conformer = guest_mol.GetConformer(conf_idx)
+                guest_conf = np.array(conformer.GetPositions(), dtype=np.float64)
+                guest_conf = guest_conf/10 # convert to md_units
+                guest_conf = recenter(guest_conf, conf_com)
+                x0 = np.concatenate([host_conf, guest_conf])       # combined geometry
+
+                combined_pdb = Chem.CombineMols(Chem.MolFromPDBFile(host_pdb_file, removeHs=False), init_mol)
+                combined_pdb_str = StringIO(Chem.MolToPDBBlock(combined_pdb))
+                out_file = os.path.join(args.out_dir, "epoch_"+str(epoch)+"_insertion_deletion_"+host_name+"_conf_"+str(conf_idx)+".pdb")
+                writer = PDBWriter(combined_pdb_str, out_file)
+
+                v0 = np.zeros_like(x0)
+
+                parent_conn, child_conn = Pipe()
+                parent_conns.append(parent_conn)
+                # writer can be None if we don't care about vis
+                all_args.append([x0, v0, conf_idx % num_gpus, writer, child_conn])
+
+            processes = []
+
+            for arg in all_args:
+                p = Process(target=sim.run_forward_and_backward, args=arg)
+                p.daemon = True
+                processes.append(p)
+                p.start()
+
+            all_du_dls = []
+            for pc in parent_conns:
+                du_dls = pc.recv()
+                all_du_dls.append(du_dls)           
+                pc.close()
+
+            # terminate all the processes
+            for p in processes:
+                p.terminate()
```

Comment:
I have to do p.terminate in the test batch instead of p.join because the process doesn't join properly.

Reply:
hm why doesn't it join properly? that's a little concerning